zapcore/write_syncer.go

```diff
@@ -75,6 +78,98 @@ func (s *lockedWriteSyncer) Sync() error {
 	return err
 }
 
+type bufferWriterSyncer struct {
+	sync.Mutex
+	bufferWriter *bufio.Writer
+	ticker       *time.Ticker
+}
+
+// defaultBufferSize sizes the buffer associated with each WriterSync.
+const defaultBufferSize = 256 * 1024
+
+// defaultFlushInterval means the default flush interval
+const defaultFlushInterval = 30 * time.Second
+
+// CloseFunc should be called when the caller exits to clean up buffers.
+type CloseFunc func() error
+
+// Buffer wraps a WriteSyncer in a buffer to improve performance,
+// if bufferSize = 0, we set it to defaultBufferSize
+// if flushInterval = 0, we set it to defaultFlushInterval
+func Buffer(ws WriteSyncer, bufferSize int, flushInterval time.Duration) (WriteSyncer, CloseFunc) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	closefunc := func() error {
+		cancel()
+		return ws.Sync()
+	}
+
+	if _, ok := ws.(*bufferWriterSyncer); ok {
+		// no need to layer on another buffer
+		return ws, closefunc
+	}
+
+	if bufferSize == 0 {
+		bufferSize = defaultBufferSize
+	}
+
+	if flushInterval == 0 {
+		flushInterval = defaultFlushInterval
+	}
+
+	ticker := time.NewTicker(flushInterval)
+
+	ws = &bufferWriterSyncer{
+		bufferWriter: bufio.NewWriterSize(ws, bufferSize),
+		ticker:       ticker,
+	}
+
+	// flush buffer every interval
+	// we do not need to exit this goroutine until closefunc called explicitly
+	go func() {
+		for {
```

Comment:
do we have any test validating that this is called in a loop?

Reply:
Nice question, test case added.